pkg/service/application_cache.go

```diff
@@ -0,0 +1,70 @@
+package service
+
+import (
+	"context"
+	"reflect"
+	"time"
+
+	"github.com/pyroscope-io/pyroscope/pkg/storage"
+)
+
+type ApplicationWriter interface {
+	CreateOrUpdate(ctx context.Context, application storage.ApplicationMetadata) error
```

Comment:
Nit: This is the same interface as `storage.ApplicationMetadataSaver`. It may make sense to use the same one... not sure.